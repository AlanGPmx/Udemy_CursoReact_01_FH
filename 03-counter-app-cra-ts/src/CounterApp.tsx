import { useState } from 'react';
import PropTypes from 'prop-types';
import { IPropsCounterApp } from './interfaces/global.interface';

export const CounterApp = (prop: IPropsCounterApp) => {
	const { valueInitial } = prop;
	const [count, setCount] = useState(valueInitial);

	const handleAdd = () => setCount(count + 1);

	const handleSubstract = () => {
		setCount((counter) => counter - 1);
	};

	const handleReset = () => {
		setCount(valueInitial);
	};

	return (
		<>
			<h1>CounterAPP</h1>
			<h2>{count}</h2>
			<button data-testid="btnIncrementarUno" onClick={handleAdd}>
				+1
			</button>{' '}
			<button data-testid="btnDecrementarUno" onClick={handleSubstract}>
				-1
			</button>
			<button aria-label="btnResetearValor" onClick={handleReset}>
				Reset
			</button>
		</>
	);
};

// Es la alternativa en JS y React al tipado que TS ofrece
CounterApp.propTypes = {
	valueInitial: PropTypes.number.isRequired,
};

//La alternativa para asignar un valor por defecto a las Props y no hacerlo en crudo en los argumentos del componente
CounterApp.defaultProps = {
	valueInitial: 0,
};
