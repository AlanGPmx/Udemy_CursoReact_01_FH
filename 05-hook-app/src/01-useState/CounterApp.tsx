import {useState} from 'react';

export const CounterApp = () => {
	const [state, setState] = useState({
		counter1: 10,
		counter2: 20,
		counter3: 30,
	});

	const {counter1, counter2, counter3} = state;

	return (
		<>
			<h2 className='text-center mb-5'>1.- Hook: useState</h2>

			<div className='row'>
				<div className='col-md-4'>
					<button
						className='btn btn-danger float-end'
						onClick={() => setState({...state, counter1: counter1 - 1})}>
						-1
					</button>
				</div>
				<div className='col-md-4 text-center'>
					<h3>Counter1: {counter1} </h3>
				</div>
				<div className='col-md-4 '>
					<button
						className='btn btn-success float-start'
						onClick={() => setState({...state, counter1: counter1 + 1})}>
						+1
					</button>
				</div>
			</div>

			<hr className='border border-primary border-1 w-50 mx-auto opacity-50' />

			<div className='row'>
				<div className='col-md-4'>
					<button
						className='btn btn-danger float-end'
						onClick={() => setState({...state, counter2: counter2 - 1})}>
						-1
					</button>
				</div>
				<div className='col-md-4 text-center'>
					<h3>Counter2: {counter2} </h3>
				</div>
				<div className='col-md-4 '>
					<button
						className='btn btn-success float-start'
						onClick={() => setState({...state, counter2: counter2 + 1})}>
						+1
					</button>
				</div>
			</div>

			<hr className='border border-primary border-1 w-50 mx-auto opacity-50' />

			<div className='row'>
				<div className='col-md-4'>
					<button
						className='btn btn-danger float-end'
						onClick={() => setState({...state, counter3: counter3 - 1})}>
						-1
					</button>
				</div>
				<div className='col-md-4 text-center'>
					<h3>Counter3: {counter3} </h3>
				</div>
				<div className='col-md-4 '>
					<button
						className='btn btn-success float-start'
						onClick={() => setState({...state, counter3: counter3 + 1})}>
						+1
					</button>
				</div>
			</div>
		</>
	);
};
