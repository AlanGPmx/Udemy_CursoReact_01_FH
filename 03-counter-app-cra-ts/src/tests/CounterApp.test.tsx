import { fireEvent, render, screen } from '@testing-library/react';
import { CounterApp } from '../CounterApp';

describe('Se realizan pruebas en el Componente <CounterAPP />', () => {
	const valueInitialTest: number = 0;
	const actualValue: number = valueInitialTest;

	test('Debe hacer match con el Snapshot', () => {
		const { container } = render(
			<CounterApp valueInitial={valueInitialTest} />
		);
		expect(container).toMatchSnapshot();
	});

	test('Debe mostrar el valor inicial de 0', () => {
		render(<CounterApp valueInitial={valueInitialTest} />);
		const counter = screen.getByRole('heading', { level: 2 });
		expect(screen.getByText(0)).toBeTruthy();
		expect(counter.innerHTML).toBe(`${valueInitialTest}`);
	});

	test('Debe incrementar el contador en +1', () => {
		render(<CounterApp valueInitial={valueInitialTest} />);
		const button = screen.getByTestId('btnIncrementarUno');
		fireEvent.click(button);

		const counter = screen.getByRole('heading', { level: 2 });
		expect(counter.innerHTML).toBeTruthy();
		expect(counter.innerHTML).toBe(`${actualValue + 1}`);
	});

	test('Debe decrementar el contador en -1', () => {
		render(<CounterApp valueInitial={valueInitialTest} />);
		const button = screen.getByTestId('btnDecrementarUno');
		fireEvent.click(button);

		const counter = screen.getByRole('heading', { level: 2 });
		expect(counter.innerHTML).toBeTruthy();
		expect(counter.innerHTML).toBe(`${actualValue - 1}`);
	});

	test('Debe resetear el contador al valor incial', () => {
		render(<CounterApp valueInitial={99} />);
		const valueInitialComponent = screen.getByRole('heading', { level: 2 });

		const btnIncrementar = screen.getByTestId('btnIncrementarUno');
		fireEvent.click(btnIncrementar);

		const btnResetear = screen.getByRole('button', {
			name: 'btnResetearValor',
		});
		fireEvent.click(btnResetear);

		const counter = screen.getByRole('heading', { level: 2 });

		expect(counter.innerHTML).toBeTruthy();
		expect(counter.innerHTML).toBe(`${valueInitialComponent.innerHTML}`);
	});
});
