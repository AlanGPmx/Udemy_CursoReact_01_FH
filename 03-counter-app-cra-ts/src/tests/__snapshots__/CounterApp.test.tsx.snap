// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Se realizan pruebas en el Componente <CounterAPP /> Debe hacer match con el Snapshot 1`] = `
<div>
  <h1>
    CounterAPP
  </h1>
  <h2>
    0
  </h2>
  <button
    data-testid="btnIncrementarUno"
  >
    +1
  </button>
   
  <button
    data-testid="btnDecrementarUno"
  >
    -1
  </button>
  <button
    aria-label="btnResetearValor"
  >
    Reset
  </button>
</div>
`;
